// vike/dist/cjs/node/vite/plugins/pluginNonRunnableDev.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pluginNonRunnableDev = pluginNonRunnableDev;
const utils_js_1 = require("../utils.js");
const retrievePageAssetsDev_js_1 = require("../../runtime/renderPage/getPageAssets/retrievePageAssetsDev.js");
const getViteConfigRuntime_js_1 = require("../shared/getViteConfigRuntime.js");
const getMagicString_js_1 = require("../shared/getMagicString.js");
(0, utils_js_1.assertIsNotProductionRuntime)();
function getViteRpcFunctions(viteDevServer) {
    return {
        async transformIndexHtmlRPC(html) {
            return await viteDevServer.transformIndexHtml('/', html);
        },
        async retrievePageAssetsDevRPC(clientDependencies, clientEntries) {
            return await (0, retrievePageAssetsDev_js_1.retrievePageAssetsDev)(viteDevServer, clientDependencies, clientEntries);
        },
        async getViteConfigRuntimeRPC() {
            return (0, getViteConfigRuntime_js_1.getViteConfigRuntime)(viteDevServer.config);
        },
    };
}
function pluginNonRunnableDev() {
    const distFileIsNonRunnableDev = (0, utils_js_1.requireResolveDistFile)('dist/esm/utils/isNonRunnableDev.js');
    const distFileGlobalContext = (0, utils_js_1.requireResolveDistFile)('dist/esm/node/runtime/globalContext.js');
    let config;
    return {
        name: 'vike:pluginNonRunnableDev',
        configureServer(viteDevServer) {
            (0, utils_js_1.createViteRPC)(viteDevServer, getViteRpcFunctions);
        },
        configResolved(config_) {
            config = config_;
        },
        transform(code, id) {
            if (!config._isDev)
                return;
            if (id !== distFileIsNonRunnableDev && id !== distFileGlobalContext)
                return;
            if ((0, utils_js_1.isRunnableDevEnvironment)(this.environment))
                return;
            const { magicString, getMagicStringResult } = (0, getMagicString_js_1.getMagicString)(code, id);
            if (id === distFileIsNonRunnableDev) {
                magicString.replaceAll('__VIKE__IS_NON_RUNNABLE_DEV', JSON.stringify(true));
            }
            if (id === distFileGlobalContext) {
                magicString.replaceAll('__VIKE__DYNAMIC_IMPORT', 'import');
            }
            return getMagicStringResult();
        },
    };
}
