// vike/dist/cjs/node/shared/virtualFileId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.virtualFileIdGlobalEntryClientCR = exports.virtualFileIdGlobalEntryClientSR = exports.virtualFileIdGlobalEntryServer = void 0;
exports.parseVirtualFileId = parseVirtualFileId;
exports.generateVirtualFileId = generateVirtualFileId;
const extractAssetsQuery_js_1 = require("./extractAssetsQuery.js");
const utils_js_1 = require("./utils.js");
(0, utils_js_1.assertIsNotBrowser)();
// Global entries
const virtualFileIdGlobalEntryServer = 
//
'virtual:vike:global-entry:server';
exports.virtualFileIdGlobalEntryServer = virtualFileIdGlobalEntryServer;
const virtualFileIdGlobalEntryClientSR = 
//
'virtual:vike:global-entry:client:server-routing';
exports.virtualFileIdGlobalEntryClientSR = virtualFileIdGlobalEntryClientSR;
const virtualFileIdGlobalEntryClientCR = 
//
'virtual:vike:global-entry:client:client-routing';
exports.virtualFileIdGlobalEntryClientCR = virtualFileIdGlobalEntryClientCR;
// Page entries
const virtualFileIdPageEntryClient = 
//
'virtual:vike:page-entry:client:'; // ${pageId}
const virtualFileIdPageEntryServer = 
//
'virtual:vike:page-entry:server:'; //  ${pageId}
// Virtual ID prefixes
const virtualFileIdPageEntryPrefix = 
//
'virtual:vike:page-entry:';
const virtualFileIdGlobalEntryPrefix = 
//
'virtual:vike:global-entry:';
const virtualFileIdGlobalEntries = [
    virtualFileIdGlobalEntryServer,
    virtualFileIdGlobalEntryClientCR,
    virtualFileIdGlobalEntryClientSR,
];
(0, utils_js_1.assert)(virtualFileIdGlobalEntries.every((v) => 
//
v.startsWith(virtualFileIdGlobalEntryPrefix)));
(0, utils_js_1.assert)([virtualFileIdPageEntryClient, virtualFileIdPageEntryServer].every((v) => 
//
v.startsWith(virtualFileIdPageEntryPrefix)));
function parseVirtualFileId(id) {
    id = (0, utils_js_1.removeVirtualFileIdPrefix)(id);
    if (!id.startsWith(virtualFileIdGlobalEntryPrefix) && !id.startsWith(virtualFileIdPageEntryPrefix))
        return false;
    // Global entry
    if (id.includes(virtualFileIdGlobalEntryPrefix)) {
        (0, utils_js_1.assert)(virtualFileIdGlobalEntries.includes(id));
        const isForClientSide = id !== virtualFileIdGlobalEntryServer;
        const isClientRouting = id === virtualFileIdGlobalEntryClientCR;
        return {
            type: 'global-entry',
            isForClientSide,
            isClientRouting,
        };
    }
    // Page entry
    if (id.includes(virtualFileIdPageEntryPrefix)) {
        (0, utils_js_1.assert)(id.startsWith(virtualFileIdPageEntryPrefix));
        const idOriginal = id;
        id = (0, extractAssetsQuery_js_1.extractAssetsRemoveQuery)(id);
        const isExtractAssets = idOriginal !== id;
        if (id.startsWith(virtualFileIdPageEntryClient)) {
            (0, utils_js_1.assert)(isExtractAssets === false);
            return {
                type: 'page-entry',
                pageId: id.slice(virtualFileIdPageEntryClient.length),
                isForClientSide: true,
                isExtractAssets,
            };
        }
        if (id.startsWith(virtualFileIdPageEntryServer)) {
            return {
                type: 'page-entry',
                pageId: id.slice(virtualFileIdPageEntryServer.length),
                isForClientSide: false,
                isExtractAssets,
            };
        }
        (0, utils_js_1.assert)(false);
    }
    return false;
}
function generateVirtualFileId(args) {
    if (args.type === 'global-entry') {
        const { isForClientSide, isClientRouting } = args;
        (0, utils_js_1.assert)(typeof isClientRouting === 'boolean');
        if (!isForClientSide) {
            return virtualFileIdGlobalEntryServer;
        }
        else if (isClientRouting) {
            return virtualFileIdGlobalEntryClientCR;
        }
        else {
            return virtualFileIdGlobalEntryClientSR;
        }
    }
    if (args.type === 'page-entry') {
        const { pageId, isForClientSide } = args;
        (0, utils_js_1.assert)(typeof pageId === 'string');
        const id = `${isForClientSide ? virtualFileIdPageEntryClient : virtualFileIdPageEntryServer}${pageId}`;
        return id;
    }
    (0, utils_js_1.assert)(false);
}
