// vike/dist/cjs/node/vite/shared/isViteServerSide.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isViteServerSide = isViteServerSide;
exports.isViteClientSide = isViteClientSide;
exports.isViteServerSide_viteEnvOptional = isViteServerSide_viteEnvOptional;
exports.isViteServerSide_onlySsrEnv = isViteServerSide_onlySsrEnv;
exports.isViteServerSide_extraSafe = isViteServerSide_extraSafe;
const assert_js_1 = require("../../../utils/assert.js");
function isViteServerSide_impl(configGlobal, viteEnv) {
    (0, assert_js_1.assert)(!('consumer' in configGlobal)); // make sure configGlobal isn't viteEnv.config
    const debug = {
        viteEnvIsUndefined: !viteEnv,
        viteEnvName: viteEnv?.name ?? null,
        viteEnvConsumer: viteEnv?.config.consumer ?? null,
        configEnvBuildSsr: viteEnv?.config.build?.ssr ?? null,
        configGlobalBuildSsr: configGlobal.build?.ssr ?? null,
    };
    if (!viteEnv) {
        const isServerSide = getBuildSsrValue(configGlobal.build?.ssr);
        (0, assert_js_1.assert)(typeof isServerSide === 'boolean', debug);
        return isServerSide;
    }
    else {
        const isServerSide1 = !viteEnv.config.consumer ? null : viteEnv.config.consumer !== 'client';
        const isServerSide2 = getBuildSsrValue(viteEnv.config.build?.ssr);
        const isServerSide3 = viteEnv.name === 'ssr' ? true : viteEnv.name === 'client' ? false : null;
        const isServerSide = isServerSide1 ?? isServerSide2;
        (0, assert_js_1.assert)(isServerSide === isServerSide1 || isServerSide1 === null, debug);
        (0, assert_js_1.assert)(isServerSide === isServerSide2 || isServerSide2 === null, debug);
        (0, assert_js_1.assert)(isServerSide === isServerSide3 || isServerSide3 === null, debug);
        (0, assert_js_1.assert)(isServerSide !== null);
        return isServerSide;
    }
}
function getBuildSsrValue(buildSsr) {
    if (buildSsr === undefined)
        return null;
    (0, assert_js_1.assert)(typeof buildSsr === 'boolean' || typeof buildSsr === 'string');
    return !!buildSsr;
}
function isViteServerSide(configGlobal, viteEnv) {
    return isViteServerSide_impl(configGlobal, viteEnv);
}
function isViteServerSide_viteEnvOptional(configGlobal, viteEnv) {
    return isViteServerSide_impl(configGlobal, viteEnv);
}
function isViteClientSide(configGlobal, viteEnv) {
    return !isViteServerSide(configGlobal, viteEnv);
}
// Only `ssr` env: for example don't include `vercel_edge` nor `vercel_node`.
function isViteServerSide_onlySsrEnv(configGlobal, viteEnv) {
    return viteEnv.name ? viteEnv.name === 'ssr' : isViteServerSide(configGlobal, viteEnv);
}
// Vite is quite messy about setting config.build.ssr — for security purposes, we use an extra safe implementation with lots of assertions, which is needed for the .client.js and .server.js guarantee.
function isViteServerSide_extraSafe(config, viteEnv, options) {
    const isServerSide = isViteServerSide(config, viteEnv);
    const debug = {
        isServerSide,
        configCommand: config.command,
        configBuildSsr: getBuildSsrValue(config.build.ssr),
        optionsIsUndefined: options === undefined,
        optionsSsr: options?.ssr ?? null,
    };
    (0, assert_js_1.assert)(options, debug);
    (0, assert_js_1.assert)(typeof options.ssr === 'boolean', debug);
    (0, assert_js_1.assert)(options.ssr === isServerSide, debug);
    return isServerSide;
}
