// vike/dist/cjs/node/runtime/renderPage/handlePageContextRequestUrl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handlePageContextRequestUrl = handlePageContextRequestUrl;
const getPageContextRequestUrl_js_1 = require("../../../shared/getPageContextRequestUrl.js");
const modifyUrl_js_1 = require("../../../shared/modifyUrl.js");
const utils_js_1 = require("../utils.js");
// See also shared/getPageContextRequestUrl.ts
function handlePageContextRequestUrl(url) {
    const urlParsed = (0, utils_js_1.parseUrl)(url, utils_js_1.baseServer);
    if (!isMatch(urlParsed)) {
        return {
            isPageContextJsonRequest: false,
            urlWithoutPageContextRequestSuffix: url,
        };
    }
    else {
        const { urlWithoutPageContextRequestSuffix, searchVikeArgs } = processUrl(urlParsed, url);
        const previousUrl = parseSearchVikeArgs(searchVikeArgs);
        return {
            /* TO-DO/soon/once: pass & use previousUrl
            isPageContextJsonRequest: { previousUrl },
            /*/
            isPageContextJsonRequest: true,
            //*/
            urlWithoutPageContextRequestSuffix,
        };
    }
}
function isMatch(urlParsed) {
    const { pathnameOriginal, pathname } = urlParsed;
    (0, utils_js_1.assert)(pathname.endsWith(getPageContextRequestUrl_js_1.pageContextJsonFileExtension) === pathnameOriginal.endsWith(getPageContextRequestUrl_js_1.pageContextJsonFileExtension));
    return pathname.endsWith(getPageContextRequestUrl_js_1.pageContextJsonFileExtension);
}
function processUrl(urlParsed, url) {
    // We cannot use `urlParsed.pathname` because it would break the `urlParsed.pathnameOriginal` value of subsequent `parseUrl()` calls.
    const { pathnameOriginal, search } = urlParsed;
    (0, utils_js_1.assert)(getPageContextRequestUrl_js_1.doNotCreateExtraDirectory === false);
    const urlSuffix = `/index${getPageContextRequestUrl_js_1.pageContextJsonFileExtension}`;
    (0, utils_js_1.assert)(pathnameOriginal.endsWith(urlSuffix), { url });
    let pathnameModified = (0, utils_js_1.slice)(pathnameOriginal, 0, -1 * urlSuffix.length);
    if (pathnameModified === '')
        pathnameModified = '/';
    const searchVikeArgs = search?._vike;
    const urlWithoutPageContextRequestSuffix = (0, modifyUrl_js_1.modifyUrl)(url, {
        pathname: pathnameModified,
        search: {
            _vike: searchVikeArgs ? null : undefined,
        },
    });
    return {
        searchVikeArgs,
        urlWithoutPageContextRequestSuffix,
    };
}
function parseSearchVikeArgs(searchVikeArgs) {
    const args = {
        previousUrl: null,
    };
    if (searchVikeArgs) {
        const parsed = JSON.parse(searchVikeArgs);
        (0, utils_js_1.assert)((0, utils_js_1.isObject)(parsed));
        if ('previousUrl' in parsed) {
            (0, utils_js_1.assert)((0, utils_js_1.hasProp)(parsed, 'previousUrl', 'string'));
            args.previousUrl = parsed.previousUrl;
        }
    }
    return args;
}
