// vike/dist/cjs/node/runtime/renderPage/loadPageConfigsLazyServerSide.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadPageConfigsLazyServerSide = loadPageConfigsLazyServerSide;
const getPageFiles_js_1 = require("../../../shared/getPageFiles.js");
const resolveVikeConfigPublic_js_1 = require("../../../shared/page-configs/resolveVikeConfigPublic.js");
const analyzePageClientSide_js_1 = require("../../../shared/getPageFiles/analyzePageClientSide.js");
const utils_js_1 = require("../utils.js");
const getPageAssets_js_1 = require("./getPageAssets.js");
const findPageConfig_js_1 = require("../../../shared/page-configs/findPageConfig.js");
const analyzePage_js_1 = require("./analyzePage.js");
const loadAndParseVirtualFilePageEntry_js_1 = require("../../../shared/page-configs/loadAndParseVirtualFilePageEntry.js");
const execHookServer_js_1 = require("./execHookServer.js");
const getCacheControl_js_1 = require("./getCacheControl.js");
async function loadPageConfigsLazyServerSide(pageContext) {
    (0, utils_js_1.objectAssign)(pageContext, {
        _pageConfig: (0, findPageConfig_js_1.findPageConfig)(pageContext._globalContext._pageConfigs, pageContext.pageId),
    });
    // Load the page's + files
    (0, utils_js_1.updateType)(pageContext, await loadPageUserFiles(pageContext));
    // Resolve new computed pageContext properties
    (0, utils_js_1.updateType)(pageContext, await resolvePageContext(pageContext));
    // Execute +onCreatePageContext
    await (0, execHookServer_js_1.execHookServer)('onCreatePageContext', pageContext);
    return pageContext;
}
async function resolvePageContext(pageContext) {
    const { isHtmlOnly, clientEntries, clientDependencies } = (0, analyzePage_js_1.analyzePage)(pageContext);
    const passToClient = [];
    const errMsgSuffix = ' should be an array of strings.';
    const isV1Design = !!pageContext._pageConfig;
    if (!isV1Design) {
        pageContext.exportsAll.passToClient?.forEach((e) => {
            (0, utils_js_1.assertUsage)((0, utils_js_1.hasProp)(e, 'exportValue', 'string[]'), `${e.exportSource}${errMsgSuffix}`);
            passToClient.push(...e.exportValue);
        });
    }
    else {
        pageContext.from.configsCumulative.passToClient?.values.forEach((v) => {
            const { value, definedAt } = v;
            const errMsg = `+passToClient value defined at ${definedAt}${errMsgSuffix}`;
            //*/ TO-DO/next-major-release: remove the passToClient once setting from the public API
            (0, utils_js_1.assertUsage)((0, utils_js_1.isArray)(value), `+passToClient value defined at ${definedAt} should be an array`);
            const valS = value.map((el) => {
                if ((0, utils_js_1.isObject)(el)) {
                    (0, utils_js_1.assertWarning)(!('once' in el), 'The passToClient once setting is deprecated and no longer has any effect. Instead, see the upcoming .once.js suffix (see https://github.com/vikejs/vike/issues/2566 for more information).', { onlyOnce: true });
                    (0, utils_js_1.assertUsage)((0, utils_js_1.hasProp)(el, 'prop', 'string'), errMsg);
                    return el.prop;
                }
                (0, utils_js_1.assertUsage)(typeof el === 'string', errMsg);
                return el;
            });
            /*/
            assertUsage(isArrayOfStrings(value), errMsg)
            //*/
            passToClient.push(...valS);
        });
    }
    (0, utils_js_1.objectAssign)(pageContext, {
        Page: pageContext.exports.Page,
        _isHtmlOnly: isHtmlOnly,
        _passToClient: passToClient,
        headersResponse: resolveHeadersResponse(pageContext),
    });
    (0, utils_js_1.objectAssign)(pageContext, {
        __getPageAssets: async () => {
            if ('_pageAssets' in pageContext) {
                return pageContext._pageAssets;
            }
            else {
                const pageAssets = await (0, getPageAssets_js_1.getPageAssets)(pageContext, clientDependencies, clientEntries);
                (0, utils_js_1.objectAssign)(pageContext, { _pageAssets: pageAssets });
                return pageContext._pageAssets;
            }
        },
    });
    // TO-DO/next-major-release: remove
    Object.assign(pageContext, {
        _getPageAssets: async () => {
            (0, utils_js_1.assertWarning)(false, 'pageContext._getPageAssets() deprecated, see https://vike.dev/preloading', {
                onlyOnce: true,
                showStackTrace: true,
            });
            const pageAssetsOldFormat = [];
            (await pageContext.__getPageAssets()).forEach((p) => {
                if (p.assetType === 'script' && p.isEntry) {
                    pageAssetsOldFormat.push({
                        src: p.src,
                        preloadType: null,
                        assetType: 'script',
                        mediaType: p.mediaType,
                    });
                }
                pageAssetsOldFormat.push({
                    src: p.src,
                    preloadType: p.assetType,
                    assetType: p.assetType === 'style' ? 'style' : 'preload',
                    mediaType: p.mediaType,
                });
            });
            return pageAssetsOldFormat;
        },
    });
    return pageContext;
}
async function loadPageUserFiles(pageContext) {
    const [{ configPublicPageLazy }] = await Promise.all([
        loadPageUserFiles_v1Design(pageContext),
        (0, analyzePageClientSide_js_1.analyzePageClientSideInit)(pageContext._globalContext._pageFilesAll, pageContext.pageId, {
            sharedPageFilesAlreadyLoaded: true,
        }),
    ]);
    (0, utils_js_1.objectAssign)(pageContext, configPublicPageLazy);
    return pageContext;
}
async function loadPageUserFiles_v1Design(pageContext) {
    const pageFilesServerSide = (0, getPageFiles_js_1.getPageFilesServerSide)(pageContext._pageFilesAll, pageContext.pageId);
    const isDev = !pageContext._globalContext._isProduction;
    const pageConfigLoaded = !pageContext._pageConfig
        ? null
        : await (0, loadAndParseVirtualFilePageEntry_js_1.loadAndParseVirtualFilePageEntry)(pageContext._pageConfig, isDev);
    await Promise.all(pageFilesServerSide.map((p) => p.loadFile?.()));
    const configPublicPageLazy = (0, resolveVikeConfigPublic_js_1.resolveVikeConfigPublicPageLazyLoaded)(pageFilesServerSide, pageConfigLoaded, pageContext._globalContext._pageConfigGlobal);
    return {
        configPublicPageLazy,
        pageFilesLoaded: pageFilesServerSide,
    };
}
function resolveHeadersResponse(pageContext) {
    const headersResponse = mergeHeaders(pageContext.config.headersResponse);
    if (!headersResponse.get('Cache-Control')) {
        const cacheControl = (0, getCacheControl_js_1.getCacheControl)(pageContext.pageId, pageContext._globalContext._pageConfigs);
        if (cacheControl)
            headersResponse.set('Cache-Control', cacheControl);
    }
    return headersResponse;
}
function mergeHeaders(headersList = []) {
    const headersMerged = new Headers();
    headersList.forEach((headers) => {
        new Headers(headers).forEach((value, key) => {
            headersMerged.append(key, value);
        });
    });
    return headersMerged;
}
