// vike/dist/cjs/node/runtime/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderErrorPage = exports.createPageRenderer = exports._injectAssets = exports.createDevMiddleware = exports.getGlobalContextAsync = exports.getGlobalContextSync = exports.getGlobalContext = exports.version = exports.stampPipe = exports.pipeStream = exports.pipeNodeStream = exports.pipeWebStream = exports.dangerouslySkipEscape = exports.escapeInject = exports.renderPage = void 0;
const isBrowser_js_1 = require("../../utils/isBrowser.js");
const assert_js_1 = require("../../utils/assert.js");
(0, assert_js_1.assert)(!(0, isBrowser_js_1.isBrowser)());
var renderPage_js_1 = require("./renderPage.js");
Object.defineProperty(exports, "renderPage", { enumerable: true, get: function () { return renderPage_js_1.renderPage; } });
var renderHtml_js_1 = require("./html/renderHtml.js");
Object.defineProperty(exports, "escapeInject", { enumerable: true, get: function () { return renderHtml_js_1.escapeInject; } });
Object.defineProperty(exports, "dangerouslySkipEscape", { enumerable: true, get: function () { return renderHtml_js_1.dangerouslySkipEscape; } });
var stream_js_1 = require("./html/stream.js");
Object.defineProperty(exports, "pipeWebStream", { enumerable: true, get: function () { return stream_js_1.pipeWebStream; } });
Object.defineProperty(exports, "pipeNodeStream", { enumerable: true, get: function () { return stream_js_1.pipeNodeStream; } });
Object.defineProperty(exports, "pipeStream", { enumerable: true, get: function () { return stream_js_1.pipeStream; } });
Object.defineProperty(exports, "stampPipe", { enumerable: true, get: function () { return stream_js_1.stampPipe; } });
var utils_js_1 = require("./utils.js");
Object.defineProperty(exports, "version", { enumerable: true, get: function () { return utils_js_1.PROJECT_VERSION; } });
var globalContext_js_1 = require("./globalContext.js");
Object.defineProperty(exports, "getGlobalContext", { enumerable: true, get: function () { return globalContext_js_1.getGlobalContext; } });
Object.defineProperty(exports, "getGlobalContextSync", { enumerable: true, get: function () { return globalContext_js_1.getGlobalContextSync; } });
Object.defineProperty(exports, "getGlobalContextAsync", { enumerable: true, get: function () { return globalContext_js_1.getGlobalContextAsync; } });
var index_js_1 = require("../runtime-dev/index.js");
Object.defineProperty(exports, "createDevMiddleware", { enumerable: true, get: function () { return index_js_1.createDevMiddleware; } });
// TO-DO/next-major-release: remove
// Deprecated exports
__exportStar(require("../../types/index-dreprecated.js"), exports);
var injectAssets__public_js_1 = require("./html/injectAssets/injectAssets__public.js");
Object.defineProperty(exports, "_injectAssets", { enumerable: true, get: function () { return injectAssets__public_js_1.injectAssets__public; } });
var createPageRenderer_js_1 = require("../createPageRenderer.js");
Object.defineProperty(exports, "createPageRenderer", { enumerable: true, get: function () { return createPageRenderer_js_1.createPageRenderer; } });
const utils_js_2 = require("./utils.js");
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
const abort_js_1 = require("../../shared/route/abort.js");
/** @deprecated
 * Replace:
 *   ```
 *   import { RenderErrorPage } from 'vike'
 *   ```
 * With:
 *   ```
 *   import { render } from 'vike/abort'
 *   ```
 *
 * See https://vike.dev/render
 */
const RenderErrorPage = (...args) => {
    (0, utils_js_2.assertWarning)(false, [
        'Replace:',
        picocolors_1.default.red("  import { RenderErrorPage } from 'vike'"),
        'With:',
        picocolors_1.default.green("  import { render } from 'vike/abort'"),
        'See https://vike.dev/render',
    ].join('\n'), { onlyOnce: true, showStackTrace: true });
    return (0, abort_js_1.RenderErrorPage)(...args);
};
exports.RenderErrorPage = RenderErrorPage;
