// vike/dist/cjs/shared/page-configs/resolveVikeConfigPublic.js
"use strict";
// TO-DO/soon: rename PageConfig names
// - Use `Internal` suffix, i.e. {Page,Global}ConfigInternal
//   - While keeping {Page,Global}ConfigPublic or remove Public suffix and rename it to {Page,Global}Config ?
// - rename EagerLoaded EagerlyLoaded
// - remove `LazyLoaded` suffix
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveVikeConfigPublicGlobal = resolveVikeConfigPublicGlobal;
exports.resolveVikeConfigPublicPageEagerLoaded = resolveVikeConfigPublicPageEagerLoaded;
exports.resolveVikeConfigPublicPageLazyLoaded = resolveVikeConfigPublicPageLazyLoaded;
const assert_exports_old_design_js_1 = require("../getPageFiles/assert_exports_old_design.js");
const getConfigDefinedAt_js_1 = require("./getConfigDefinedAt.js");
const helpers_js_1 = require("./helpers.js");
const utils_js_1 = require("../utils.js");
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
function resolveVikeConfigPublicPageEagerLoaded(pageConfigGlobalValues, pageConfig, pageConfigValues) {
    const vikeConfigPublicPage_ = resolveVikeConfigPublic_base({ pageConfigGlobalValues, pageConfigValues });
    const vikeConfigPublicPage = getPublicCopy(vikeConfigPublicPage_);
    let page;
    if (!pageConfig.isErrorPage) {
        const route = vikeConfigPublicPage.config.route ?? pageConfig.routeFilesystem.routeString;
        page = {
            ...vikeConfigPublicPage,
            route,
        };
    }
    else {
        page = {
            ...vikeConfigPublicPage,
            isErrorPage: true,
        };
    }
    return [pageConfig.pageId, page];
}
function getPublicCopy(vikeConfigPublic) {
    return {
        config: vikeConfigPublic.config,
        _source: vikeConfigPublic.source,
        _sources: vikeConfigPublic.sources,
        _from: vikeConfigPublic.from,
    };
}
function resolveVikeConfigPublic_base({ pageConfigGlobalValues, pageConfigValues, }) {
    const configValues = { ...pageConfigGlobalValues, ...pageConfigValues };
    return resolveVikeConfigPublic_V1Design({ configValues });
}
function resolveVikeConfigPublicGlobal({ pageConfigGlobalValues, }) {
    const vikeConfigPublicGlobal = resolveVikeConfigPublic_V1Design({ configValues: pageConfigGlobalValues });
    return getPublicCopy(vikeConfigPublicGlobal);
}
function resolveVikeConfigPublicPageLazyLoaded(pageFiles, // V0.4 design
pageConfig, // V1 design
pageConfigGlobal) {
    const config = {};
    const configEntries = {}; // TO-DO/next-major-release: remove
    const exportsAll = {}; // TO-DO/next-major-release: remove
    // V0.4 design
    // TO-DO/next-major-release: remove
    pageFiles.forEach((pageFile) => {
        const exportValues = getExportValues(pageFile);
        exportValues.forEach(({ exportName, exportValue, isFromDefaultExport }) => {
            (0, utils_js_1.assert)(exportName !== 'default');
            exportsAll[exportName] = exportsAll[exportName] ?? [];
            exportsAll[exportName].push({
                exportValue,
                exportSource: `${pageFile.filePath} > ${isFromDefaultExport ? `\`export default { ${exportName} }\`` : `\`export { ${exportName} }\``}`,
                filePath: pageFile.filePath,
                _filePath: pageFile.filePath, // TO-DO/next-major-release: remove
                _fileType: pageFile.fileType,
                _isFromDefaultExport: isFromDefaultExport,
            });
        });
    });
    let source;
    let sources;
    let from;
    if (pageConfig) {
        const res = resolveVikeConfigPublic_base({
            pageConfigGlobalValues: pageConfigGlobal.configValues,
            pageConfigValues: pageConfig.configValues,
        });
        source = res.source;
        sources = res.sources;
        from = res.from;
        Object.assign(config, res.config);
        Object.assign(configEntries, res.configEntries);
        Object.assign(exportsAll, res.exportsAll);
    }
    else {
        source = {};
        sources = {};
        from = {
            configsStandard: {},
            configsCumulative: {},
            configsComputed: {},
        };
    }
    const pageExports = {};
    const exports = {};
    Object.entries(exportsAll).forEach(([exportName, values]) => {
        values.forEach(({ exportValue, _fileType, _isFromDefaultExport }) => {
            exports[exportName] = exports[exportName] ?? exportValue;
            // Legacy pageContext.pageExports
            if (_fileType === '.page' && !_isFromDefaultExport) {
                if (!(exportName in pageExports)) {
                    pageExports[exportName] = exportValue;
                }
            }
        });
    });
    (0, utils_js_1.assert)(!('default' in exports));
    (0, utils_js_1.assert)(!('default' in exportsAll));
    const configPublicPageLazy = {
        config: config,
        from,
        source,
        sources,
        // TO-DO/soon/flat-pageContext: deprecate every prop below
        configEntries,
        exports,
        exportsAll,
    };
    // TO-DO/next-major-release: remove
    (0, utils_js_1.objectDefineProperty)(configPublicPageLazy, 'pageExports', {
        get: () => {
            // We only show the warning in Node.js because when using Client Routing Vue integration uses `Object.assign(pageContextReactive, pageContext)` which will wrongully trigger the warning. There is no cross-browser way to catch whether the property accessor was initiated by an `Object.assign()` call.
            if (!(0, utils_js_1.isBrowser)()) {
                (0, utils_js_1.assertWarning)(false, 'pageContext.pageExports is outdated, use pageContext.exports instead', {
                    onlyOnce: true,
                    showStackTrace: true,
                });
            }
            return pageExports;
        },
        enumerable: false,
        configurable: true,
    });
    return configPublicPageLazy;
}
// V1 design
function resolveVikeConfigPublic_V1Design(pageConfig) {
    const config = {};
    const configEntries = {};
    const exportsAll = {};
    const source = {};
    const sources = {};
    const from = {
        configsStandard: {},
        configsCumulative: {},
        configsComputed: {},
    };
    const addSrc = (src, configName) => {
        source[configName] = src;
        sources[configName] ?? (sources[configName] = []);
        sources[configName].push(src);
    };
    const addLegacy = (configName, value, definedAtData) => {
        const configValueFilePathToShowToUser = (0, helpers_js_1.getConfigValueFilePathToShowToUser)(definedAtData);
        const configDefinedAt = (0, getConfigDefinedAt_js_1.getConfigDefinedAtOptional)('Config', configName, definedAtData);
        configEntries[configName] = configEntries[configName] ?? [];
        configEntries[configName].push({
            configValue: value,
            configDefinedAt,
            configDefinedByFile: configValueFilePathToShowToUser,
        });
        // TO-DO/next-major-release: remove
        const exportName = configName;
        exportsAll[exportName] = exportsAll[exportName] ?? [];
        exportsAll[exportName].push({
            exportValue: value,
            exportSource: configDefinedAt,
            filePath: configValueFilePathToShowToUser,
            _filePath: configValueFilePathToShowToUser,
            _fileType: null,
            _isFromDefaultExport: null,
        });
    };
    Object.entries(pageConfig.configValues).forEach(([configName, configValue]) => {
        const { value } = configValue;
        config[configName] = config[configName] ?? value;
        if (configValue.type === 'standard') {
            const src = {
                type: 'configsStandard',
                value: configValue.value,
                definedAt: (0, getConfigDefinedAt_js_1.getDefinedAtString)(configValue.definedAtData, configName),
            };
            addSrc(src, configName);
            from.configsStandard[configName] = src;
            addLegacy(configName, value, configValue.definedAtData);
        }
        if (configValue.type === 'cumulative') {
            const src = {
                type: 'configsCumulative',
                definedAt: (0, getConfigDefinedAt_js_1.getDefinedAtString)(configValue.definedAtData, configName),
                values: configValue.value.map((value, i) => {
                    const definedAtFile = configValue.definedAtData[i];
                    (0, utils_js_1.assert)(definedAtFile);
                    const definedAt = (0, getConfigDefinedAt_js_1.getDefinedAtString)(definedAtFile, configName);
                    addLegacy(configName, value, definedAtFile);
                    return {
                        value,
                        definedAt,
                    };
                }),
            };
            addSrc(src, configName);
            from.configsCumulative[configName] = src;
        }
        if (configValue.type === 'computed') {
            const src = {
                type: 'configsComputed',
                definedAt: 'Vike', // Vike currently doesn't support user-land computed configs => computed configs are always defined by Vike => there isn't any file path to show.
                value: configValue.value,
            };
            addSrc(src, configName);
            from.configsComputed[configName] = src;
            addLegacy(configName, value, configValue.definedAtData);
        }
    });
    return {
        config: config,
        configEntries,
        exportsAll,
        source,
        sources,
        from,
    };
}
function getExportValues(pageFile) {
    const { filePath, fileExports } = pageFile;
    (0, utils_js_1.assert)(fileExports); // assume pageFile.loadFile() was called
    (0, utils_js_1.assert)((0, utils_js_1.isScriptFile)(filePath));
    const exportValues = [];
    Object.entries(fileExports)
        .sort((0, utils_js_1.makeLast)(([exportName]) => exportName === 'default')) // `export { bla }` should override `export default { bla }`
        .forEach(([exportName, exportValue]) => {
        let isFromDefaultExport = exportName === 'default';
        if (isFromDefaultExport) {
            if ((0, utils_js_1.isTemplateFile)(filePath)) {
                exportName = 'Page';
            }
            else {
                (0, utils_js_1.assertUsage)((0, utils_js_1.isObject)(exportValue), `The ${picocolors_1.default.cyan('export default')} of ${filePath} should be an object.`);
                Object.entries(exportValue).forEach(([defaultExportName, defaultExportValue]) => {
                    (0, assert_exports_old_design_js_1.assertDefaultExports)(defaultExportName, filePath);
                    exportValues.push({
                        exportName: defaultExportName,
                        exportValue: defaultExportValue,
                        isFromDefaultExport,
                    });
                });
                return;
            }
        }
        exportValues.push({
            exportName,
            exportValue,
            isFromDefaultExport,
        });
    });
    exportValues.forEach(({ exportName, isFromDefaultExport }) => {
        (0, utils_js_1.assert)(!(isFromDefaultExport && assert_exports_old_design_js_1.forbiddenDefaultExports.includes(exportName)));
    });
    return exportValues;
}
