// vike/dist/cjs/node/vite/plugins/pluginBuild/pluginDistPackageJsonFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pluginDistPackageJsonFile = pluginDistPackageJsonFile;
const rollupIsEsm_js_1 = require("../../shared/rollupIsEsm.js");
const isViteServerSide_js_1 = require("../../shared/isViteServerSide.js");
function pluginDistPackageJsonFile() {
    let config;
    return {
        name: 'vike:build:pluginDistPackageJsonFile',
        apply: 'build',
        configResolved(config_) {
            config = config_;
        },
        generateBundle(options, bundle) {
            if (!(0, isViteServerSide_js_1.isViteServerSide)(config, this.environment))
                return;
            const isEsm = (0, rollupIsEsm_js_1.rollupIsEsm)(options);
            const fileName = 'package.json';
            if (bundle[fileName])
                return; // E.g. already generated by Telefunc / vike
            this.emitFile({
                fileName,
                type: 'asset',
                source: getPackageJsonContent(isEsm),
            });
        },
    };
}
function getPackageJsonContent(isEsm) {
    if (isEsm) {
        return `{ "type": "module" }`;
    }
    else {
        return `{ "type": "commonjs" }`;
    }
}
