// vike/dist/esm/node/vite/plugins/pluginNonRunnableDev.js
export { pluginNonRunnableDev };
import { createViteRPC, assertIsNotProductionRuntime, requireResolveDistFile, isRunnableDevEnvironment, } from '../utils.js';
import { retrievePageAssetsDev } from '../../runtime/renderPage/getPageAssets/retrievePageAssetsDev.js';
import { getViteConfigRuntime } from '../shared/getViteConfigRuntime.js';
import { getMagicString } from '../shared/getMagicString.js';
assertIsNotProductionRuntime();
function getViteRpcFunctions(viteDevServer) {
    return {
        async transformIndexHtmlRPC(html) {
            return await viteDevServer.transformIndexHtml('/', html);
        },
        async retrievePageAssetsDevRPC(clientDependencies, clientEntries) {
            return await retrievePageAssetsDev(viteDevServer, clientDependencies, clientEntries);
        },
        async getViteConfigRuntimeRPC() {
            return getViteConfigRuntime(viteDevServer.config);
        },
    };
}
function pluginNonRunnableDev() {
    const distFileIsNonRunnableDev = requireResolveDistFile('dist/esm/utils/isNonRunnableDev.js');
    const distFileGlobalContext = requireResolveDistFile('dist/esm/node/runtime/globalContext.js');
    let config;
    return {
        name: 'vike:pluginNonRunnableDev',
        configureServer(viteDevServer) {
            createViteRPC(viteDevServer, getViteRpcFunctions);
        },
        configResolved(config_) {
            config = config_;
        },
        transform(code, id) {
            if (!config._isDev)
                return;
            if (id !== distFileIsNonRunnableDev && id !== distFileGlobalContext)
                return;
            if (isRunnableDevEnvironment(this.environment))
                return;
            const { magicString, getMagicStringResult } = getMagicString(code, id);
            if (id === distFileIsNonRunnableDev) {
                magicString.replaceAll('__VIKE__IS_NON_RUNNABLE_DEV', JSON.stringify(true));
            }
            if (id === distFileGlobalContext) {
                magicString.replaceAll('__VIKE__DYNAMIC_IMPORT', 'import');
            }
            return getMagicStringResult();
        },
    };
}
